src/execution.rs

```diff
@@ -114,6 +123,97 @@ pub trait Report {
     fn start(&mut self, task_name: String);
     fn report(&mut self, result: &CompletedTask);
     fn done(&mut self);
+
+    fn stage(&mut self, full_name: &Vec<String>, stage_rep: StageReport) {
+        let mut full_name = full_name.clone();
+        match stage_rep {
+            StageReport::Start { stage_name } => {
+                full_name.push(stage_name);
+                self.start(full_name.join("::"));
+            }
+            StageReport::Stop {
+                stage_name,
+                status,
+                duration,
+            } => {
+                full_name.push(stage_name);
+                let completed_task = CompletedTask {
+                    full_name,
+                    duration,
+                    stdout: Vec::new(),
+                    stderr: Vec::new(),
+                    status: Status::from(status),
+                };
+                self.report(&completed_task);
+            }
+        }
+    }
+}
+
+pub struct StageReportSender {
+    sender: pipe::Sender,
+    // Keeping the stages that have been "started" around makes it easy
+    // to add a duration later on.
+    started_stages: BTreeMap<String, SystemTime>,
```

Comment:
Why it needs to be a btree map? I assume stages should start in order, so an array might be more appropriate.

Reply:
It doesn't. I removed it.